### DISCUSSION TITLE
CMS field overrides with _schema.json are really fun yet powerful

### DISCUSSION THREAD

**Discussion by jamestagal** [plentico/plenti#266] [Show and tell]
Hi all,
I just wanted to show how you could add a `list` item or a `paragraph` to your content when they are defined as components (that aren't part of default content). See screencast of a list item example. 

https://user-images.githubusercontent.com/6309595/212783937-d6a924eb-cc90-48c5-90f5-068054e06893.mov

1. In your `_components` folder create another folder and call it `item`. Then, in the `_defaults.json` file there, add a default list item as follows:

```json
{
	"item": "Another list item"
}
```
2. Next in your `_schema.json` file in the content when you want this item to be applied in, (in my case it was for the `courses` content) add the field name/s and `item` as the option specifying the `type` as a `component`.
```json
{
  "list1": {
    "type": "component",
    "options": [
      "item"
    ]
  },
  "list2": {
    "type": "component",
    "options": [
      "item"
    ]
  }
} 
```
3. Make sure you put your `item/s` into an array in your data source, as follows:
```json
"list1": [
        {
        "item": "Lorem ipsum dolor sit amet consectetur adipisicing elit."
        }
 ],
"list2": [
        {
        "item": "Lorem ipsum dolor sit amet consectetur adipisicing elit."
        }
 ],
```
4. Finally loop over the `item` in your html template as follows:
```html
        <ul>
          {#each list1 as l}
          <li>{@html l.item}</li>
        {/each}
        </ul>
```
Hope that is helpful. 
Thanks the guidance on this one @jimafisk 
Ben

**Comment by jimafisk** [comment]
Thanks for the guide @jamestagal!

**Comment by dindinet** [comment]
Hi @jamestagal and @jimafisk

Firstly brilliant project. Great sollution to Git backed CMS and especially previewing. I am a fan of the NetlifyCMS but the Preview templating is a pain in the ***. The NetlifyCMS 'workflow' using branches is very helpful for teams that require reviews. 

I tried to implement adding new faq items with the bigspring template using CMS field overrides. 

					{
						"question": "New faq Q?",
						"answer": "new faq A"
					}

I tried to follow @jamestagal example but while not breaking the CMS I could not for the life of me get the 'Add New Faq item' button to surface. My problem I think was where to place all the files especially the _schema.json. I did not follow step 4. I figured the faq.svelte code would work as is.

Do either of you have an idea of what would be required to get the result I am after?

Regards
DaveL

**Reply by dindinet** [reply]
Woaah! 
I found the [youtube channel for Plenti CMS](https://www.youtube.com/playlist?list=PLbWvcwWtuDm3vNn5ANzgVjyL1YOICT0jE)
Found episodes 25, 26 and 27 useful. Will try overiding the faq component item again and report back.

**Reply by jimafisk** [reply]
Thanks for checking out the project @dindinet! 

> I am a fan of the NetlifyCMS but the Preview templating is a pain in the ***.

Same! We built Plenti because we needed [better previews](https://github.com/decaporg/decap-cms/issues/3911) and wanted closer coupling to our data source to [reduce duplicating the field config](https://github.com/decaporg/decap-cms/issues/322) in the SSG and CMS.

> The NetlifyCMS 'workflow' using branches is very helpful for teams that require reviews.

Definitely, this is something that we'd like to add in the future.

> I tried to implement adding new faq items with the bigspring template using CMS field overrides.

This was the first Plenti theme I built, so there are still places where it hasn't been updated to show off all the new features. The new features can certainly be added, just haven't done so yet.

> I could not for the life of me get the 'Add New Faq item' button to surface. My problem I think was where to place all the files especially the _schema.json.

There are a few things required to extend the CMS to allow adding new FAQs:
1. Add a `content/_components/` folder
2. Create a `content/_components/faq/` folder for the FAQ component
3. Add a `content/_components/faq/_defaults.json` file (can just copy the FAQ component JSON from `content/pages/faq.json`)
4. Add a `content/_components/faq/_schema.json` file with the following contents:
```json
{
    "fields.items": {
        "type": "components",
        "options": [
            "faq_item"
        ]
    }
}
```
5. Create a `content/_components/faq_item/` folder for the FAQ Item component
6. Add a `content/_components/faq_item/_defaults.json` file (can just copy the Items component JSON from `content/_components/faq/_defaults.json`)

That sounds more complicated than it is in practice. I just pushed a commit to the BigSpring repo to demonstrate: https://github.com/plenti-themes/bigspring/commit/1f91a32ccd04bb462f123133888c38e44c82af85

> I found the [youtube channel for Plenti CMS](https://www.youtube.com/playlist?list=PLbWvcwWtuDm3vNn5ANzgVjyL1YOICT0jE)

Glad you found the channel, hopefully it helps some. I still change the API every once in a while so sometimes the videos get out of date (feel free to add a comment to the video if you find something out of date). Don't hesitate to reach out with questions!

**Reply by jamestagal** [reply]
Yes nice example @jimafisk  and great to see you are exploring Plenti's CMS features @dindinet. 

Just a note. The only difference with my example and the one Jim illustrated is that I created that `item` as a component from a  ```content``` type in Plenti. So in my case my field override is placed in the ```_schema.json``` file of that ```content``` type not in a component as a component override.  Hope that further clarifies the difference.  Happy days! 😃

**Reply by dindinet** [reply]
Hi @jimafisk @jamestagal 

I followed the instructions from @jimafisk for adding faq_items. That worked. I then tried to implement something similar for the blurbs on the privacy-policy page and the terms-conditions page. However, the interface offered up faq_item as an add an item option in the cms for the two pages.
![Screenshot from 2023-04-30 16-55-42](https://user-images.githubusercontent.com/2939526/235363158-e0e680d7-c99d-4d3b-8775-a7ae0df7aec0.png)
I failed to get that to work, I'm still swingining at windmills with this stuff. I haven't built up a mental model of what plenti is doing in the build. 

@jimafisk Might it be feasable to add a true or false **add_new** key to any content array and have the cms discovery auto add an
 '**+ item**' button. (similar to the NetlifyCMS list widget). If this were possible it would be really helpful for content that is deeply nested.

		{
			"name": "faq",
                        "add_new": true
			"fields": {
				"items": [
					{
						"question": "Will updates also be free?",
						"answer": "Lorem, link ipsum dolor sit amet consectetur adipisicing elit."
					},
					{
						"question": "Discounts for students and Non Profit Organizations?",
						"answer": "Lorem, link ipsum dolor sit amet consectetur adipisicing elit. "
					}
				]

			}
		}

Regards
Dave L

**Reply by jimafisk** [reply]
> However, the interface offered up faq_item as an add an item option in the cms for the two pages.

The way we figure out which schema to apply is by comparing the key structure of the JSON on the page to the key structure of the defaults for the component (so a `_defaults.json` file is required). This lets us track components without needing separate mapping config, but it requires unique key structure. If both the `faq` and `blurbs` components used the same items convention, then Plenti wouldn't be able to distinguish between them. However, that doesn't appear to be the case, so adding the components (https://github.com/plenti-themes/bigspring/commit/794ca6f4b6c126db561dfa0bbff52e28f92ecf8d) seems to be working on my end:

<details>
<summary>See Blurb Component screenshot</summary>

![blurb](https://user-images.githubusercontent.com/5913244/235380692-5cc5849c-d951-4ac7-9e1c-771e01151ea2.png)


</details>

That said, this theme was developed a while ago and I actually think this would be better as a `wysiwyg` field widget. I've actually gone and converted that (https://github.com/plenti-themes/bigspring/commit/549c4066696725972e431d3fa8bddffa4d826cca):

<details>
<summary>See WYSIWYG screenshot</summary>

![wysiwyg](https://user-images.githubusercontent.com/5913244/235381560-c58e0045-47c6-4e1e-9685-69c72d082d7f.png)


</details>

If you want these updates, just pull the latest theme changes into your project: `plenti theme update bigspring`

> Might it be feasable to add a true or false add_new key to any content array and have the cms discovery auto add an
'+ item' button.

Unfortunately that violates our principal of unstructured JSON. There are no required keys to allows folks to pull from a variety of content sources and APIs that they might not necessarily have control over. I often use a component-based architecture with the `name` / `fields` convention, but it's possible to use Plenti variety of dissimilar ways based on user preference.

**Reply by dindinet** [reply]
Hi @jimafisk 

I had another play with this. To get the CMS to offer an appropriate set of components for different pages I ended up creating
a folder for each page in the content folder. This was to allow for each page using it's own schema  and defaults file.

I thought I had seen somewhere that by using a naming convention like _schema_contenttype.json and _defaults_contenttype.json
I need not have created the separate folders. Did I dream that?

Starting to get more of a handle on it and still enthusing about plenti. :)

Dave L

**Reply by jimafisk** [reply]
> To get the CMS to offer an appropriate set of components for different pages I ended up creating
a folder for each page in the content folder

If that fits your project requirements, that's totally valid! The way I typically structure pages is I use a `pages` type with a schema that defines the component library available to all pages. Each page typically uses a different set of components on the actual page, but the component library works as a type of "page builder" that editors can use add/remove or move content around.

> I thought I had seen somewhere that by using a naming convention like _schema_contenttype.json and _defaults_contenttype.json

Yes that does exist, it's called a "single" content type. You can define a schema for a single type by creating a JSON file with `_schema_` prepended to the front of it (like `_schema_mytype.json`). There are no _`defaults` for single types though, since defaults are used to add more instances of the same type and these are one-off by definition. There are some more details here: https://plenti.co/docs/types

> Starting to get more of a handle on it and still enthusing about plenti. :)

That's great, thanks for sharing your progress!
